Encryption docs update

aditighag opened the pull request:
This is a follow-up to #14924.

- Adds more troubleshooting steps.
- Add steps for IPAM EMI mode.
- Match installation and validation steps.

Fixes: https://github.com/cilium/cilium/pull/15695/commits/b6ec84ccb400368b65647aff53a79316e8dd547a

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -260,28 +345,79 @@ Troubleshooting
      .. code:: bash
 
-        ip xfrm s
+        ip -s s xfrm s
```
Is this meant to be `ip -s xfrm s`?

```
$ ip -s s xfrm s
Error: argument "xfrm" is wrong: unknown
```

(also elsewhere)

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -260,28 +345,79 @@ Troubleshooting
+
+        tc filter show dev lxc6c63af7b7f62 ingress
+        filter protocol all pref 1 bpf chain 0
+        filter protocol all pref 1 bpf chain 0 handle 0x1 bpf_lxc.o:[from-container] direct-action not_in_hw id 1494 tag 4921d192be8621ca jited
```
Do we need this particular piece for programs on the lxc devices? Cilium is pretty thoroughly broken if these are not configured.

aditighag replied:
Yup, agreed.

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -231,13 +284,45 @@ Troubleshooting
+
+                ip route list table 200
+                10.93.0.0/16 dev cilium_host mtu 9001
+                local 10.135.0.0/16 dev eth0 proto 50 scope host
```
Is there a meaningful difference between GKE / ENI here? The device may be different, but even in the GKE case you might not be running with tunnel mode so I don't think that particular example is 100% accurate.

aditighag replied:
I didn't find the best way to converge both the tabs with references in the following sections. What do you recommend?

joestringer replied:
What I had in mind is to first tell the user that the device is either `cilium_host` or `cilium_vxlan` or whatever, then you provide one example. The rest of the info seems to be exactly the same across both environments.

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -200,6 +201,58 @@ used by Cilium. The key variable is a uint8 with value between 0-16 and should
+Run a ``bash`` shell in one of the Cilium pods with ``kubectl -n kube-system
+exec -ti <cilium pod> -- bash`` and execute the following commands:
+
+1. Install tcpdump
```
I don't think we should be encouraging people to randomly install stuff in their Cilium containers and parse tcpdump output just to check whether Cilium is working correctly.

This should be possible via `cilium status`.

joestringer replied:
^ Alternate suggestion before we implement visibility in `cilium status` is to just deploy the connectivity check. Those should fail if ipsec is misconfigured right? (I guess it doesn't double-check for ESP packets though, hmm..)

joestringer replied:
I guess we're already doing this in the encryption guide, so this can be a non-blocking comment. I'd like for this to be improved in Cilium.

joestringer replied:
though if it's already in the guide, is this duplicating the "validate the setup" step? Can we better organize this to avoid such duplication?

aditighag replied:
Yes, I'm aware of the duplication. :)

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -200,6 +201,58 @@ used by Cilium. The key variable is a uint8 with value between 0-16 and should
+      --set ipam.mode=eni \\
+      --set tunnel=disabled \\
+      --set encryption.enabled=true \\
+      --set encryption.nodeEncryption=false
```
I'll defer to @jrfastab here but from what I understand, the user must also specify `--ipv4-pod-subnets` to cover all subnets that AWS may allocate IPs from within the cluster. Note this is not currently exposed in the helm charts, so users would need to either use `extraArgs` or we'd need to introduce dedicated helm options for this setting.

qmonnet commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -255,6 +283,41 @@ func retrieveIfIndexFromMAC(mac mac.MAC, mtu int) (index int, err error) {
+func retrieveIfaceIdxFromMAC(mac string) (int, error) {
+	iface, err := retrieveIfaceFromMAC(mac)
+	if err != nil {
+		err = fmt.Errorf("failed to get iface index with MAC %w", err)
```
```suggestion
		err = fmt.Errorf("failed to get iface index with MAC: %w", err)
```

aditighag replied:
I'll push another commit to fix these formatting issues.

qmonnet commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -223,6 +223,17 @@ func SetupRules(from, to *net.IPNet, mac string, ifaceNum int) error {
+func RetrieveIfaceNameFromMAC(mac string) (string, error) {
+	iface, err := retrieveIfaceFromMAC(mac)
+	if err != nil {
+		err = fmt.Errorf("failed to get iface name with MAC %w", err)
```
```suggestion
		err = fmt.Errorf("failed to get iface name with MAC: %w", err)
```

qmonnet commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -200,6 +201,58 @@ used by Cilium. The key variable is a uint8 with value between 0-16 and should
+Encryption with IPAM ENI mode
+==============================
+
+In order to enable encryption on EKS with the IPAM ENI mode, deploy cilium via
```
```suggestion
In order to enable encryption on EKS with the IPAM ENI mode, deploy Cilium via
```

qmonnet commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -200,6 +201,58 @@ used by Cilium. The key variable is a uint8 with value between 0-16 and should
+
+1. Install tcpdump
+
+.. code:: bash
```
```suggestion
.. code-block:: shell-session
```
(Possibly with `$`s at the beginning of the commands)
(Same suggestion further down in the doc)

qmonnet commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -260,28 +345,79 @@ Troubleshooting
+   * BPF program to encrypt and decrypt traffic:
+
+    Check if the BPF programs to encrypt and decrypt traffic are attached on
+    on correct interfaces.
```
```suggestion
    correct interfaces.
```

aditighag replied:
Maybe you meant to suggest "to"?

qmonnet reviewed (approved):
Few suggestions on minor nitpicks, looks good to me besides Joe's concerns.

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -157,25 +183,50 @@ exec -ti <cilium pod> -- bash`` and execute the following commands:
+         .. code-block:: shell-session
+
+            $ tcpdump -n -i cilium_vxlan
+            tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
```
I found it a bit confusing that these tabs are named "GKE", but earlier in this document there are two sets of GKE instructions, one with native mode and one with tunneling. Maybe these GKE example tabs should be just `Tunneling mode` instead?

jrfastab replied:
should we drop the tunnel mode or at least discourage it somehow. Its not a very optimal setup for us and AFAIK we don't encourage tunnel+GKE at the moment.

joestringer replied:
I think the tunnel piece was just coming from the previous version of the doc. The main question here is how we ensure that regardless of the /other/ configuration that the user wants to use, it's clear which encryption options they should add on top.

Longer-term, the new CLI will solve this problem.

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -224,26 +275,62 @@ Troubleshooting
+
+                .. note::
+                    Skip this step if ``endpoint-routes`` is enabled.
+
```
`endpoint-routes` is always enabled in IPAM ENI mode, so the step can always be ignored.

I don't think we need the note at all here?

joestringer commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -224,26 +275,62 @@ Troubleshooting
+               local 10.135.0.0/16 dev eth0 proto 50 scope host
 
-     .. code:: bash
+            Content of per-ENI routing table:
```
Oh, and the indentation seems to be doing funny things here when rendered, could you double-check that?

aditighag replied:
I did run `make render-docs` before pushing my changes. I'll double check.

joestringer replied:
`make render-docs` will not pick up on odd indentation for what it's worth. In general sphinx seems to like 3-space indents (basically the next set of text is indented to the same indentation as the "code" part here).

I think the only way to really pick up on these is to render the docs and observe the generated output pages.

jrfastab commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -115,6 +116,31 @@ traffic) can be enabled as well. See :ref:`node_to_node_encryption` below.
+      --set ipam.mode=eni \\
+      --set tunnel=disabled \\
+      --set encryption.enabled=true \\
+      --set encryption.nodeEncryption=false
```
Additionally need the ipv4-pod-subnet option or its not going to work reliably for >2 nodes.

aditighag replied:
I'm not sure I fully understand that option. Do you want to fix that in a follow-up PR? If the documentation is quite straight forward, I'm happy to include it in this PR as well.

jrfastab commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -115,6 +116,31 @@ traffic) can be enabled as well. See :ref:`node_to_node_encryption` below.
+      --set encryption.nodeEncryption=false
+
+By default, this enables routing via the ``cilium_host`` interface.
+Specifying encryption interfaces with ``--set encryption.interface=ethX`` is not supported.
```
We fixed this right? I needed it to make EKS work in my setup when auto-detection did not pick the correct device.

aditighag replied:
This PR was sent prior to your EKS encryption fixes, I'll update this line.

aditighag replied:
Well, we haven't exactly fixed this, have we? The user specified interface config can cause traffic disruption if `BPF_NETWORK` prog that does decryption is not attached to that interface.

jrfastab commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -115,6 +116,31 @@ traffic) can be enabled as well. See :ref:`node_to_node_encryption` below.
+.. note::
+
+    For IPAM ENI mode, the user-provided interface may lead to issues where traffic
+    traffic may not get decrypted.
```
huh? If this is true lets block it at config time to ensure system doesn't come up, but really we should fix it.

jrfastab replied:
I need some more details on why 'traffic may not get decrypted.'

aditighag replied:
Per Slack conversation, you said you were planning to file an issue to document the edge cases? I can link to the issue here.

joestringer replied:
I agree, saying "this might not work" without providing any further details (how to identify the problem, how to fix) only serves to confuse users.

We should clearly define what we expect to happen. If there's a common pitfall, ideally we would mitigate it in software, but that's not possible, we should add a troubleshooting subsection that defines how you identify an issue like this and how you fix it.

joestringer replied:
^ One way to tackle this is something like saying:

It is not required to specify `--interfaces` (or whatever the option is called). If however you do specify this flag, it will also configure the set of devices that decryption is applied on. As a result, if this option is misconfigured then it can cause connectivity issues between nodes as encrypted traffic cannot be decrypted at the destination. See the troubleshooting section (link here) for more details to identify and resolve this issue.

jrfastab commented on Documentation/gettingstarted/encryption.rst:
```diff
@@ -157,25 +183,50 @@ exec -ti <cilium pod> -- bash`` and execute the following commands:
+        .. code-block:: shell-session
+
+            $ tcpdump -n -i eth1
+            tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
```
can we compress this output some. I don't think the mac addresses are needed.

jrfastab reviewed (changes requested):
Needs a couple fixes and added a few comments inline.

aditighag commented on Documentation/gettingstarted/encryption-ipsec.rst:
```diff
@@ -265,25 +283,65 @@ Troubleshooting
+
+   * BPF program to decrypt traffic:
+
+    Check if the BPF programs to decrypt traffic is attached to relevant interfaces.
```
How does this read? AFAIR, we now attach the program to all except reserved interfaces (e.g., veth, etc)? Should we mention `eth0` (default) and `encryption-interface`?

joestringer replied:
I think so. But yeah the reader's question then becomes "Which interfaces are relevant?" . Maybe 'all network-facing interfaces or alternatively all interfaces matching the configuration of `--encrypt-interface`'?

Also one nit, I think it should be "programs ... are" or "program ... is":
```suggestion
    Check if the BPF program to decrypt traffic is attached to relevant interfaces.
```

joestringer replied:
oh and I wouldn't suggest mentioning `eth0` specifically here since some common distros like Ubuntu won't create such an interface (as it uses the new alphabet soup naming scheme). It's OK as an example, but we shouldn't treat `eth0` somehow specially or create an impression that there has to be a device with that name.

joestringer commented on Documentation/gettingstarted/encryption-ipsec.rst:
```diff
@@ -161,17 +161,19 @@ following commands:
 2. Check that traffic is encrypted. In the example below, this can be verified
-   by the fact that packets carry the IP Encapsulating Security Payload (ESP):
+   by the fact that packets carry the IP Encapsulating Security Payload (ESP).
+   In the example below, ``eth0`` is essentially the interface used for pod-to-pod
```
nit, unnecessary word:
```suggestion
   In the example below, ``eth0`` is the interface used for pod-to-pod
```

joestringer reviewed (changes requested):
Looks fine to me, just a couple of nits.

I'd still like for there to be a proper way to enforce encryption and validate that it's working correctly without executing into the cilium pods and installing software since that's generally not a recommended production workflow, but in the absence of such features and for a getting started guide I think this is probably OK.

aditighag committed:
docs: Improve ipsec encryption guide

- Adds more troubleshooting steps.
- Add steps for IPAM EMI mode.
- Match installation and validation steps. (Fixes: b6ec84c)

Signed-off-by: Aditi Ghag <aditi@cilium.io>

joestringer reviewed (approved):
LGTM, thanks!
